#  Data Wrangling

# Introduction

The tasks for this project are:

* Data wrangling, which consists of:
  * Gathering data
  * Assessing data
  * Cleaning data
* Storing, analyzing, and visualizing our wrangled data
* Reporting on
* 1) our data wrangling efforts
* 2) our data analyses and visualizations

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import time
import gc
import requests
import tweepy
import os
import json
import re
import warnings

plt.style.use('ggplot')
%matplotlib inline  

# Gather

In [7]:
# read csv as a Pandas DataFrame
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [8]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [9]:
# Use requests library to download tsv file
url="https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv"
response = requests.get(url)
with open('./Data/image_predictions.tsv', 'wb') as file:
    file.write(response.content)
image_predictions = pd.read_csv('./Data/image_predictions.tsv', sep='\t')


In [10]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


**Query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file.**


In [11]:
CONSUMER_KEY = "5Uur0mo4ol2kB8yhtZ1VxXS0u"
CONSUMER_SECRET = "h8E7fSpXWiMoBel7G1ZOAeu4Mgru0v0MtxH5ehYE1RKM89SiBH"
OAUTH_TOKEN = "303562412-ct9aNnU0FQR0UKJVn1i1W3Y8omqSewiQWUcRaygB"
OAUTH_TOKEN_SECRET = "D3qslrbdOU5fqTOp951kOIuZbkeTPBodnjNYoEGFR63Ft" 

In [16]:
auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(OAUTH_TOKEN, OAUTH_TOKEN_SECRET)

api = tweepy.API(auth, 
                 wait_on_rate_limit = True, 
                 wait_on_rate_limit_notify = True)

In [19]:
# List of the error tweets
error_list = []
# List of tweets
df_list = []
# Calculate the time of execution
start = time.time()

i=0
# For loop which will add each available tweet json to df_list
for tweet_id in twitter_archive['tweet_id']:
    i+=1
    try:
        tweet = api.get_status(tweet_id, tweet_mode='extended',
                               wait_on_rate_limit = True, wait_on_rate_limit_notify = True)._json 
        favorites = tweet['favorite_count'] # How many favorites the tweet had
        retweets = tweet['retweet_count'] # Count of the retweet
        user_followers = tweet['user']['followers_count'] # How many followers the user had
        user_favourites = tweet['user']['favourites_count'] # How many favorites the user had
        date_time = tweet['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)
        
    print(np.round((i/twitter_archive.shape[0])*100, 2),'% done')
# Calculate the time of excution
end = time.time()
print(end - start)

0.04 % done
0.08 % done
0.13 % done
0.17 % done
0.21 % done
0.25 % done
0.3 % done
0.34 % done
0.38 % done
0.42 % done
0.47 % done
0.51 % done
0.55 % done
0.59 % done
0.64 % done
0.68 % done
0.72 % done
0.76 % done
0.81 % done
888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
0.85 % done
0.89 % done
0.93 % done
0.98 % done
1.02 % done
1.06 % done
1.1 % done
1.15 % done
1.19 % done
1.23 % done
1.27 % done
1.32 % done
1.36 % done
1.4 % done
1.44 % done
1.49 % done
1.53 % done
1.57 % done
1.61 % done
1.66 % done
1.7 % done
1.74 % done
1.78 % done
1.83 % done
1.87 % done
1.91 % done
1.95 % done
1.99 % done
2.04 % done
2.08 % done
2.12 % done
2.16 % done
2.21 % done
2.25 % done
2.29 % done
2.33 % done
2.38 % done
2.42 % done
2.46 % done
2.5 % done
2.55 % done
2.59 % done
2.63 % done
2.67 % done
2.72 % done
2.76 % done
2.8 % done
2.84 % done
2.89 % done
2.93 % done
2.97 % done
3.01 % done
3.06 % done
3.1 % done
3.14 % done
3.18 % done
3.23 % done
3.27 % done
3.

23.26 % done
23.3 % done
23.34 % done
23.39 % done
23.43 % done
23.47 % done
23.51 % done
23.56 % done
23.6 % done
23.64 % done
23.68 % done
23.73 % done
23.77 % done
23.81 % done
23.85 % done
23.9 % done
23.94 % done
23.98 % done
24.02 % done
802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
24.07 % done
24.11 % done
24.15 % done
24.19 % done
24.24 % done
24.28 % done
24.32 % done
24.36 % done
24.41 % done
24.45 % done
24.49 % done
24.53 % done
24.58 % done
24.62 % done
24.66 % done
24.7 % done
24.75 % done
24.79 % done
24.83 % done
24.87 % done
24.92 % done
24.96 % done
25.0 % done
25.04 % done
25.08 % done
25.13 % done
25.17 % done
25.21 % done
25.25 % done
25.3 % done
25.34 % done
25.38 % done
25.42 % done
25.47 % done
25.51 % done
25.55 % done
25.59 % done
25.64 % done
25.68 % done
25.72 % done
25.76 % done
25.81 % done
25.85 % done
25.89 % done
25.93 % done
25.98 % done
26.02 % done
26.06 % done
26.1 % done
26.15 % done
26.19 % done
26.23 % done
26.

48.3 % done
48.34 % done
48.39 % done
48.43 % done
48.47 % done
48.51 % done
48.56 % done
48.6 % done
48.64 % done
48.68 % done
48.73 % done
48.77 % done
48.81 % done
48.85 % done
48.9 % done
48.94 % done
48.98 % done
49.02 % done
49.07 % done
49.11 % done
49.15 % done
49.19 % done
49.24 % done
49.28 % done
49.32 % done
49.36 % done
49.41 % done
49.45 % done
49.49 % done
49.53 % done
49.58 % done
49.62 % done
49.66 % done
49.7 % done
49.75 % done
49.79 % done
49.83 % done
49.87 % done
49.92 % done
49.96 % done
50.0 % done
50.04 % done
50.08 % done
50.13 % done
50.17 % done
50.21 % done
50.25 % done
50.3 % done
50.34 % done
50.38 % done
50.42 % done
50.47 % done
50.51 % done
50.55 % done
50.59 % done
50.64 % done
50.68 % done
50.72 % done
50.76 % done
50.81 % done
50.85 % done
50.89 % done
50.93 % done
50.98 % done
51.02 % done
51.06 % done
51.1 % done
51.15 % done
51.19 % done
51.23 % done
51.27 % done
51.32 % done
51.36 % done
51.4 % done
51.44 % done
51.49 % done
51.53 % done
51.57 %

Rate limit reached. Sleeping for: 83


68.97 % done
69.02 % done
69.06 % done
69.1 % done
69.14 % done
69.19 % done
69.23 % done
69.27 % done
69.31 % done
69.35 % done
69.4 % done
69.44 % done
69.48 % done
69.52 % done
69.57 % done
69.61 % done
69.65 % done
69.69 % done
69.74 % done
69.78 % done
69.82 % done
69.86 % done
69.91 % done
69.95 % done
69.99 % done
70.03 % done
70.08 % done
70.12 % done
70.16 % done
70.2 % done
70.25 % done
70.29 % done
70.33 % done
70.37 % done
70.42 % done
70.46 % done
70.5 % done
70.54 % done
70.59 % done
70.63 % done
70.67 % done
70.71 % done
70.76 % done
70.8 % done
70.84 % done
70.88 % done
70.93 % done
70.97 % done
71.01 % done
71.05 % done
71.1 % done
71.14 % done
71.18 % done
71.22 % done
71.26 % done
71.31 % done
71.35 % done
71.39 % done
71.43 % done
71.48 % done
71.52 % done
71.56 % done
71.6 % done
71.65 % done
71.69 % done
71.73 % done
71.77 % done
71.82 % done
71.86 % done
71.9 % done
71.94 % done
71.99 % done
72.03 % done
72.07 % done
72.11 % done
72.16 % done
72.2 % done
72.24 % 

95.67 % done
95.71 % done
95.76 % done
95.8 % done
95.84 % done
95.88 % done
95.93 % done
95.97 % done
96.01 % done
96.05 % done
96.1 % done
96.14 % done
96.18 % done
96.22 % done
96.26 % done
96.31 % done
96.35 % done
96.39 % done
96.43 % done
96.48 % done
96.52 % done
96.56 % done
96.6 % done
96.65 % done
96.69 % done
96.73 % done
96.77 % done
96.82 % done
96.86 % done
96.9 % done
96.94 % done
96.99 % done
97.03 % done
97.07 % done
97.11 % done
97.16 % done
97.2 % done
97.24 % done
97.28 % done
97.33 % done
97.37 % done
97.41 % done
97.45 % done
97.5 % done
97.54 % done
97.58 % done
97.62 % done
97.67 % done
97.71 % done
97.75 % done
97.79 % done
97.84 % done
97.88 % done
97.92 % done
97.96 % done
98.01 % done
98.05 % done
98.09 % done
98.13 % done
98.17 % done
98.22 % done
98.26 % done
98.3 % done
98.34 % done
98.39 % done
98.43 % done
98.47 % done
98.51 % done
98.56 % done
98.6 % done
98.64 % done
98.68 % done
98.73 % done
98.77 % done
98.81 % done
98.85 % done
98.9 % done
98.94 % 

In [20]:
# lengh of the result
print("The lengh of the result", len(df_list))
# The tweet_id of the errors
print("The lengh of the errors", len(error_list))

The lengh of the result 2331
The lengh of the errors 25


From the above results:
- We reached the limit of the tweepy API three times but wait_on_rate_limit automatically wait for rate limits to re-establish and wait_on_rate_limit_notify print a notification when Tweepy is waiting.
- We could get 2344 tweet_id correctly with 12 errors
- The total time was about 3023 seconds (~ 50.5 min)

In [21]:
print("The length of the result", len(df_list))

The length of the result 2331


In [22]:
# Create DataFrames from list of dictionaries
json_tweets = pd.DataFrame(df_list, columns = ['tweet_id', 'favorites', 'retweets',
                                               'user_followers', 'user_favourites', 'date_time'])
# Save the dataFrame in file
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [23]:
# Read the saved tweet_json.txt file into a dataframe
tweet_data = pd.read_csv('tweet_json.txt', encoding = 'utf-8')
tweet_data

tweet_id  favorites  retweets  user_followers  \
0     892420643555336193      35865      7598         8822720   
1     892177421306343426      30938      5627         8822079   
2     891815181378084864      23282      3724         8822079   
3     891689557279858688      39134      7775         8822079   
4     891327558926688256      37384      8375         8822079   
...                  ...        ...       ...             ...   
2326  666049248165822465         96        40         8822137   
2327  666044226329800704        268       130         8822137   
2328  666033412701032449        111        41         8822137   
2329  666029285002620928        120        42         8822137   
2330  666020888022790149       2388       458         8822137   

      user_favourites                  date_time  
0              145809  2017-08-01 16:23:56+00:00  
1              145808  2017-08-01 00:17:27+00:00  
2              145808  2017-07-31 00:18:03+00:00  
3              145808  2017-07-30 15:58:51+00:00  
4              145808  2017-07-29 16:00:24+00:00  
...               ...                        ...  
2326           145810  2015-11-16 00:24:50+00:00  
2327           145810  2015-11-16 00:04:52+00:00  
2328           145810  2015-11-15 23:21:54+00:00  
2329           145810  2015-11-15 23:05:30+00:00  
2330           145810  2015-11-15 22:32:08+00:00  

[2331 rows x 6 columns]

In [24]:
tweet_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   tweet_id         2331 non-null   int64 
 1   favorites        2331 non-null   int64 
 2   retweets         2331 non-null   int64 
 3   user_followers   2331 non-null   int64 
 4   user_favourites  2331 non-null   int64 
 5   date_time        2331 non-null   object
dtypes: int64(5), object(1)
memory usage: 109.4+ KB


## Gather: Summary
Gathering is the first step in the data wrangling process.
- Obtaining data
   - Getting data from an existing file (twitter-archive-enhanced.csv) Reading from csv file using pandas
   - Downloading a file from the internet (image-predictions.tsv) Downloading file using requests
   - Querying an API (tweet_json.txt) Get JSON object of all the tweet_ids using Tweepy
- Importing that data into our programming environment (Jupyter Notebook)

## Assessing

In [25]:
# Print some random examples
twitter_archive.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
958   751456908746354688                    NaN                  NaN   
925   755110668769038337                    NaN                  NaN   
531   808106460588765185                    NaN                  NaN   
1429  697596423848730625                    NaN                  NaN   
1230  713411074226274305                    NaN                  NaN   
351   831322785565769729                    NaN                  NaN   
1724  680085611152338944                    NaN                  NaN   
46    883360690899218434                    NaN                  NaN   
2288  667176164155375616                    NaN                  NaN   
2310  666786068205871104                    NaN                  NaN   

                      timestamp  \
958   2016-07-08 16:44:23 +0000   
925   2016-07-18 18:43:07 +0000   
531   2016-12-12 00:29:28 +0000   
1429  2016-02-11 01:42:02 +0000   
1230  2016-03-25 17:03:49 +0000   
351   2017-02-14 02:02:51 +0000   
1724  2015-12-24 18:00:19 +0000   
46    2017-07-07 16:22:55 +0000   
2288  2015-11-19 03:02:47 +0000   
2310  2015-11-18 01:12:41 +0000   

                                                 source  \
958   <a href="http://twitter.com/download/iphone" r...   
925   <a href="http://twitter.com/download/iphone" r...   
531   <a href="http://twitter.com/download/iphone" r...   
1429  <a href="http://twitter.com/download/iphone" r...   
1230  <a href="http://twitter.com/download/iphone" r...   
351   <a href="http://twitter.com/download/iphone" r...   
1724  <a href="https://about.twitter.com/products/tw...   
46    <a href="http://twitter.com/download/iphone" r...   
2288  <a href="http://twitter.com/download/iphone" r...   
2310  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
958   Here's a pupper that's very hungry but too laz...                  NaN   
925   This is Watson. He trust falls on command. 13/...                  NaN   
531   Here we have Burke (pupper) and Dexter (doggo)...                  NaN   
1429  Here's a pupper with a piece of pizza. Two of ...                  NaN   
1230  Here we see an extremely rare Bearded Floofmal...                  NaN   
351   This is Pete. He has no eyes. Needs a guide do...                  NaN   
1724  This is by far the most coordinated series of ...                  NaN   
46    Meet Grizzwald. He may be the floofiest floofe...                  NaN   
2288  These are strange dogs. All have toupees. Long...                  NaN   
2310  Unfamiliar with this breed. Ears pointy af. Wo...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
958                        NaN                        NaN   
925                        NaN                        NaN   
531                        NaN                        NaN   
1429                       NaN                        NaN   
1230                       NaN                        NaN   
351                        NaN                        NaN   
1724                       NaN                        NaN   
46                         NaN                        NaN   
2288                       NaN                        NaN   
2310                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
958   https://twitter.com/dog_rates/status/751456908...                12   
925   https://twitter.com/dog_rates/status/755110668...                13   
531   https://twitter.com/dog_rates/status/808106460...                12   
1429  https://twitter.com/dog_rates/status/697596423...                11   
1230  https://twitter.com/dog_rates/status/713411074...                11   
351   https://twitter.com/dog_rates/status/831322785...                12   
1724  https://twitter.com/dog_rates/status/680085611...                12   
46    https://twit

In [26]:
# Assessing the data programmaticaly
twitter_archive.info()
twitter_archive.describe()
twitter_archive['rating_numerator'].value_counts()
twitter_archive['rating_denominator'].value_counts()
twitter_archive['name'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

None       745
a           55
Charlie     12
Lucy        11
Oliver      11
          ... 
Andy         1
Kanu         1
Tanner       1
Sully        1
Ginger       1
Name: name, Length: 957, dtype: int64

In [27]:
# View descriptive statistics of twitter_archive
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [28]:
image_predictions

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [29]:
image_predictions.info()
image_predictions['jpg_url'].value_counts()
image_predictions[image_predictions['jpg_url'] == 'https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   

      img_num      p1   p1_conf  p1_dog                p2  p2_conf  p2_dog  \
2070        2  basset  0.555712    True  English_springer  0.22577    True   

                               p3   p3_conf  p3_dog  
2070  German_short-haired_pointer  0.175219    True

In [30]:
# View number of entries for each source
twitter_archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [31]:
#For rating that don't follow pattern
twitter_archive[twitter_archive['rating_numerator'] > 20]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
290   838150277551247360           8.381455e+17         2.195506e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
290   2017-03-04 22:12:52 +0000   
313   2017-02-24 21:54:03 +0000   
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
188   <a href="http://twitter.com/download/iphone" r...   
189   <a href="http://twitter.com/download/iphone" r...   
290   <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/downloa

In [32]:
#unusual names
twitter_archive[twitter_archive['name'].apply(len) < 3]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706                    NaN                  NaN   
393   825876512159186944                    NaN                  NaN   
446   819015337530290176                    NaN                  NaN   
449   819004803107983360                    NaN                  NaN   
553   804026241225523202                    NaN                  NaN   
...                  ...                    ...                  ...   
2349  666051853826850816                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   

                      timestamp  \
56    2017-07-02 15:32:16 +0000   
393   2017-01-30 01:21:19 +0000   
446   2017-01-11 02:57:27 +0000   
449   2017-01-11 02:15:36 +0000   
553   2016-11-30 18:16:08 +0000   
...                         ...   
2349  2015-11-16 00:35:11 +0000   
2350  2015-11-16 00:30:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   

                                                 source  \
56    <a href="http://twitter.com/download/iphone" r...   
393   <a href="http://twitter.com/download/iphone" r...   
446   <a href="http://twitter.com/download/iphone" r...   
449   <a href="http://twitter.com/download/iphone" r...   
553   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
56    Here is a pupper approaching maximum borkdrive...                  NaN   
393   This is Mo. No one will push him around in the...                  NaN   
446   RT @dog_rates: This is Bo. He was a very good ...         8.190048e+17   
449   This is Bo. He was a very good First Doggo. 14...                  NaN   
553   This is Bo. He's going to make me cry. 13/10 p...                  NaN   
...                                                 ...                  ...   
2349  This is an odd dog. Hard on the outside but lo...                  NaN   
2350  This is a truly beautiful English Wilson Staff...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
56                         NaN                        NaN   
393                        NaN                        NaN   
446               4.196984e+09  2017-01-11 02:15:36 +0000   
449                        NaN                        NaN   
553                        NaN                        NaN   
...                        ...                        ...   
2349                       NaN                        NaN   
2350                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
56    https://twitter.com/dog_rates/status/881536004...                14   
393   https://twitter.com/dog_rates/status/825876512...                11   
446   https://twitter.com/dog_rates/status/819004803...                14   
449   https://twitt

In [33]:
#Orignal Tweets 
twitter_archive[twitter_archive['retweeted_status_id'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

## Quality
*Completeness, Validity, Accuracy, Consistency => a.k.a content issues*

**twitter_archive dataset**
- in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id should be integers/strings instead of float.
- retweeted_status_timestamp, timestamp should be datetime instead of object (string).
- The numerator and denominator columns have invalid values.
- In several columns null objects are non-null (None to NaN).
- Name column have invalid names i.e 'None', 'a', 'an' and less than 3 characters.
- We only want original ratings (no retweets) that have images.
- We may want to change this columns type (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id and tweet_id) to string because We don't want any operations on them.
- Sources difficult to read.

**image_predictions dataset**
- Missing values from images dataset (2075 rows instead of 2356)
- Some tweet_ids have the same jpg_url
- Some tweets are have 2 different tweet_id one redirect to the other (Dataset contains retweets)

**tweet_data dataset**

- This tweet_id (666020888022790149) duplicated 8 times


## Tidiness
Untidy data => a.k.a structural issues
- No need to all the informations in images dataset, (tweet_id and jpg_url what matters)
- Dog "stage" variable in four columns: doggo, floofer, pupper, puppo
- Join 'tweet_info' and 'image_predictions' to 'twitter_archive'

## Cleaning
Cleaning our data is the third step in data wrangling. It is where we will fix the quality and tidiness issues that we identified in the assess step.

In [34]:
#copy dataframes
tweet_data_clean = tweet_data.copy()
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean= image_predictions.copy()

#### Define
Add tweet_info and image_predictions to twitter_archive table.

#### Code

In [35]:
twitter_archive_clean = pd.merge(left=twitter_archive_clean,
                                 right=tweet_data_clean, left_on='tweet_id', right_on='tweet_id', how='inner')

In [36]:
twitter_archive_clean = twitter_archive_clean.merge(image_predictions_clean, on='tweet_id', how='inner')


#### Test

In [37]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2059 entries, 0 to 2058
Data columns (total 33 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2059 non-null   int64  
 1   in_reply_to_status_id       23 non-null     float64
 2   in_reply_to_user_id         23 non-null     float64
 3   timestamp                   2059 non-null   object 
 4   source                      2059 non-null   object 
 5   text                        2059 non-null   object 
 6   retweeted_status_id         72 non-null     float64
 7   retweeted_status_user_id    72 non-null     float64
 8   retweeted_status_timestamp  72 non-null     object 
 9   expanded_urls               2059 non-null   object 
 10  rating_numerator            2059 non-null   int64  
 11  rating_denominator          2059 non-null   int64  
 12  name                        2059 non-null   object 
 13  doggo                       2059 

#### Define
 Melt the 'doggo', 'floofer', 'pupper' and 'puppo' columns into one column 'dog_stage'.

#### Code

In [38]:
# Select the columns to melt and to remain
MELTS_COLUMNS = ['doggo', 'floofer', 'pupper', 'puppo']
STAY_COLUMNS = [x for x in twitter_archive_clean.columns.tolist() if x not in MELTS_COLUMNS]


# Melt the the columns into values
twitter_archive_clean = pd.melt(twitter_archive_clean, id_vars = STAY_COLUMNS, value_vars = MELTS_COLUMNS, 
                         var_name = 'stages', value_name = 'dog_stage')
# Delete column 'stages'
twitter_archive_clean = twitter_archive_clean.drop('stages', 1)

#### Test

In [39]:
print(twitter_archive_clean.dog_stage.value_counts())
print(len(twitter_archive_clean))

None       7905
pupper      221
doggo        78
puppo        24
floofer       8
Name: dog_stage, dtype: int64
8236


#### Clean
Clean rows and columns that we will not need

#### Code

In [40]:
# Delete the retweets
twitter_archive_clean = twitter_archive_clean[pd.isnull(twitter_archive_clean.retweeted_status_id)]

# Delete duplicated tweet_id
twitter_archive_clean = twitter_archive_clean.drop_duplicates()

# Delete tweets with no pictures
twitter_archive_clean = twitter_archive_clean.dropna(subset = ['jpg_url'])

# small test
len(twitter_archive_clean)

2303

In [41]:
# Delete columns related to retweet we don't need anymore
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_id', 1)
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_user_id', 1)
twitter_archive_clean = twitter_archive_clean.drop('retweeted_status_timestamp', 1)

# Delete column date_time we imported from the API, it has the same values as timestamp column
twitter_archive_clean = twitter_archive_clean.drop('date_time', 1)

# small test
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'dog_stage']

In [42]:
#Delete dog_stage duplicates
twitter_archive_clean = twitter_archive_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

#### Test

In [43]:
print(twitter_archive_clean.dog_stage.value_counts())
print(len(twitter_archive_clean))

None       1682
pupper      212
doggo        62
puppo        23
floofer       8
Name: dog_stage, dtype: int64
1987


#### Define
Get rid of image prediction columns

#### Code

In [44]:
# We will store the fisrt true algorithm with it's level of confidence
prediction_algorithm = []
confidence_level = []

# Get_prediction_confidence function:
# search the first true algorithm and append it to a list with it's level of confidence
# if flase prediction_algorthm will have a value of NaN
def get_prediction_confidence(dataframe):
    if dataframe['p1_dog'] == True:
        prediction_algorithm.append(dataframe['p1'])
        confidence_level.append(dataframe['p1_conf'])
    elif dataframe['p2_dog'] == True:
        prediction_algorithm.append(dataframe['p2'])
        confidence_level.append(dataframe['p2_conf'])
    elif dataframe['p3_dog'] == True:
        prediction_algorithm.append(dataframe['p3'])
        confidence_level.append(dataframe['p3_conf'])
    else:
        prediction_algorithm.append('NaN')
        confidence_level.append(0)

twitter_archive_clean.apply(get_prediction_confidence, axis=1)
twitter_archive_clean['prediction_algorithm'] = prediction_algorithm
twitter_archive_clean['confidence_level'] = confidence_level

#### Test

In [45]:
list(twitter_archive_clean)


['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [46]:
# Delete the columns of image prediction information
twitter_archive_clean = twitter_archive_clean.drop(['img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'], 1)

In [47]:
list(twitter_archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'favorites',
 'retweets',
 'user_followers',
 'user_favourites',
 'jpg_url',
 'dog_stage',
 'prediction_algorithm',
 'confidence_level']

In [48]:
# let's concentrate on low values.. let's dig more
twitter_archive_clean.info()
print('in_reply_to_user_id ')
print(twitter_archive_clean['in_reply_to_user_id'].value_counts())
print('source ')
print(twitter_archive_clean['source'].value_counts())
print('user_favourites ')
print(twitter_archive_clean['user_favourites'].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1987 entries, 0 to 7053
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   tweet_id               1987 non-null   int64  
 1   in_reply_to_status_id  23 non-null     float64
 2   in_reply_to_user_id    23 non-null     float64
 3   timestamp              1987 non-null   object 
 4   source                 1987 non-null   object 
 5   text                   1987 non-null   object 
 6   expanded_urls          1987 non-null   object 
 7   rating_numerator       1987 non-null   int64  
 8   rating_denominator     1987 non-null   int64  
 9   name                   1987 non-null   object 
 10  favorites              1987 non-null   int64  
 11  retweets               1987 non-null   int64  
 12  user_followers         1987 non-null   int64  
 13  user_favourites        1987 non-null   int64  
 14  jpg_url                1987 non-null   object 
 15  dog_

#### Notes
- One value in ***in_reply_to_user_id*** so we will delete the columns of reply all of them replying to @dog_rates.
- ***source** has 3 types, we will clean that column and made them clean.
- **user_favourites** has 2 values and they are close.

In [49]:
# drop the following columns 'in_reply_to_status_id', 'in_reply_to_user_id', 'user_favourites'
twitter_archive_clean = twitter_archive_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'user_favourites'], 1)

In [50]:
# Clean the content of source column
twitter_archive_clean['source'] = twitter_archive_clean['source'].apply(lambda x: re.findall(r'>(.*)<', x)[0])

In [51]:
# Test
twitter_archive_clean

tweet_id                  timestamp              source  \
0     892420643555336193  2017-08-01 16:23:56 +0000  Twitter for iPhone   
1452  679877062409191424  2015-12-24 04:11:37 +0000  Twitter for iPhone   
1451  680070545539371008  2015-12-24 17:00:27 +0000           TweetDeck   
1450  680085611152338944  2015-12-24 18:00:19 +0000           TweetDeck   
1448  680115823365742593  2015-12-24 20:00:22 +0000  Twitter for iPhone   
...                  ...                        ...                 ...   
6191  889531135344209921  2017-07-24 17:02:04 +0000  Twitter for iPhone   
6189  889665388333682689  2017-07-25 01:55:32 +0000  Twitter for iPhone   
7022  743253157753532416  2016-06-16 01:25:36 +0000  Twitter for iPhone   
6517  819952236453363712  2017-01-13 17:00:21 +0000  Twitter for iPhone   
7053  738537504001953792  2016-06-03 01:07:16 +0000  Twitter for iPhone   

                                                   text  \
0     This is Phineas. He's a mystical boy. Only eve...   
1452  Meet Penelope. She's a bacon frise. Total babe...   
1451  Say hello to Emmie. She's trapped in an orname...   
1450  This is by far the most coordinated series of ...   
1448  This is Ozzy. He woke up 2 minutes before he h...   
...                                                 ...   
6191  This is Stuart. He's sporting his favorite fan...   
6189  Here's a puppo that seems to be on the fence a...   
7022  This is Kilo. He cannot reach the snackum. Nif...   
6517  This is Oliver. He has dreams of being a servi...   
7053  This is Bayley. She fell asleep trying to esca...   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1452  https://twitter.com/dog_rates/status/679877062...                 5   
1451  https://twitter.com/dog_rates/status/680070545...                 9   
1450  https://twitter.com/dog_rates/status/680085611...                12   
1448  https://twitter.com/dog_rates/status/680115823...                 9   
...                                                 ...               ...   
6191  https://twitter.com/dog_rates/status/889531135...                13   
6189  https://twitter.com/dog_rates/status/889665388...                13   
7022  https://twitter.com/dog_rates/status/743253157...                10   
6517  https://www.gofundme.com/servicedogoliver,http...                13   
7053  https://twitter.com/dog_rates/status/738537504...                11   

      rating_denominator      name  favorites  retweets  user_followers  \
0                     10   Phineas      35865      7598         8822720   
1452                  10  Penelope       1954       637         8822124   
1451                  10     Emmie       2543       844         8822123   
1450                  10        by      12601      8811         8822123   
1448                  10      Ozzy       2671       882         8822123   
...                  ...       ...        ...       ...             ...   
6191                  10    Stuart      14079      2032         8822079   
6189                  10      None      44631      8996         8822079   
7022                  10      Kilo       4176      1208         8822108   
6517                  10    Oliver       5353      1179         8822094   
7053                  10    Bayley       5015      1507         8822108   

                                              jpg_url dog_stage  \
0     https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg      None   
1452  https://pbs.twimg.com/media/CW9olDsUsAA0XSf.jpg      None   
1451  https://pbs.twimg.com/media/CW-dU34WQAANBGy.jpg      None   
1450  https://pbs.twimg.com/media/CXAiiHUWkAIN_28.jpg      None   
1448  https://pbs.twimg.com/media/CXBBurSWMAELewi.jpg      None   
...                                               ...       ...   
6191  https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg     puppo   
6189  https://pbs.twimg.com/media/DFi579UWsA

#### Define
Fix rating numerator and denominators that are not actually ratings

#### Code

In [52]:
# View all occurences where there are more than one #/# in 'text' column
text_ratings_to_fix = twitter_archive_clean[twitter_archive_clean.text.str.contains( r"(\d+\.?\d*\/\d+\.?\d*\D+\d+\.?\d*\/\d+\.?\d*)")].text

text_ratings_to_fix

C:\Users\topp\Anaconda3\lib\site-packages\pandas\core\strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


1392    This is Darrel. He just robbed a 7/11 and is i...
1613    Two gorgeous dogs here. Little waddling dog is...
1513    Meet Tassy &amp; Bee. Tassy is pretty chill, b...
1547    These two pups just met and have instantly bon...
1216    This may be the greatest video I've ever been ...
1126    This is Socks. That water pup w the super legs...
1256    When bae says they can't go out but you see th...
1282    Meet Fynn &amp; Taco. Fynn is an all-powerful ...
1271    This is Eriq. His friend just reminded him of ...
1967    This is Kial. Kial is either wearing a cape, w...
1976    Two dogs in this one. Both are rare Jujitsu Py...
1883    Here we have Pancho and Peaches. Pancho is a C...
1920    This is Spark. He's nervous. Other dog hasn't ...
2009    These are Peruvian Feldspars. Their names are ...
2038    This is an Albanian 3 1/2 legged  Episcopalian...
1718    10/10 for dog. 7/10 for cat. 12/10 for human. ...
1820    Meet Hank and Sully. Hank is very proud of the...
1771    Meet H

In [53]:
for entry in text_ratings_to_fix:
    mask = twitter_archive_clean.text == entry
    column_name1 = 'rating_numerator'
    column_name2 = 'rating_denominator'
    twitter_archive_clean.loc[mask, column_name1] = re.findall(r"\d+\.?\d*\/\d+\.?\d*\D+(\d+\.?\d*)\/\d+\.?\d*", entry)
    twitter_archive_clean.loc[mask, column_name2] = 10

In [54]:
twitter_archive_clean[twitter_archive_clean.text.isin(text_ratings_to_fix)]

tweet_id                  timestamp              source  \
1392  682962037429899265  2016-01-01 16:30:13 +0000  Twitter for iPhone   
1613  674646392044941312  2015-12-09 17:46:48 +0000  Twitter for iPhone   
1513  677314812125323265  2015-12-17 02:30:09 +0000  Twitter for iPhone   
1547  676191832485810177  2015-12-14 00:07:50 +0000  Twitter for iPhone   
1216  695064344191721472  2016-02-04 02:00:27 +0000  Twitter for iPhone   
1126  703356393781329922  2016-02-26 23:10:06 +0000  Twitter for iPhone   
1256  691483041324204033  2016-01-25 04:49:38 +0000  Twitter for iPhone   
1282  689835978131935233  2016-01-20 15:44:48 +0000  Twitter for iPhone   
1271  690400367696297985  2016-01-22 05:07:29 +0000  Twitter for iPhone   
1967  667544320556335104  2015-11-20 03:25:43 +0000  Twitter Web Client   
1976  667491009379606528  2015-11-19 23:53:52 +0000  Twitter Web Client   
1883  669037058363662336  2015-11-24 06:17:19 +0000  Twitter for iPhone   
1920  668537837512433665  2015-11-22 21:13:35 +0000  Twitter for iPhone   
2009  666835007768551424  2015-11-18 04:27:09 +0000  Twitter for iPhone   
2038  666287406224695296  2015-11-16 16:11:11 +0000  Twitter for iPhone   
1718  672248013293752320  2015-12-03 02:56:30 +0000  Twitter for iPhone   
1820  670434127938719744  2015-11-28 02:48:46 +0000  Twitter for iPhone   
1771  671154572044468225  2015-11-30 02:31:34 +0000  Twitter for iPhone   
1221  694352839993344000  2016-02-02 02:53:12 +0000  Twitter for iPhone   
863   740373189193256964  2016-06-08 02:41:38 +0000  Twitter for iPhone   
988   716439118184652801  2016-04-03 01:36:11 +0000  Twitter for iPhone   
1004  714258258790387713  2016-03-28 01:10:13 +0000  Twitter for iPhone   
954   722974582966214656  2016-04-21 02:25:47 +0000  Twitter for iPhone   
605   777684233540206592  2016-09-19 01:42:24 +0000  Twitter for iPhone   
812   747600769478692864  2016-06-28 01:21:27 +0000  Twitter for iPhone   
5796  673295268553605120  2015-12-06 00:17:55 +0000  Twitter for iPhone   
5727  674737130913071104  2015-12-09 23:47:22 +0000  Twitter for iPhone   

                                                   text  \
1392  This is Darrel. He just robbed a 7/11 and is i...   
1613  Two gorgeous dogs here. Little waddling dog is...   
1513  Meet Tassy &amp; Bee. Tassy is pretty chill, b...   
1547  These two pups just met and have instantly bon...   
1216  This may be the greatest video I've ever been ...   
1126  This is Socks. That water pup w the super legs...   
1256  When bae says they can't go out but you see th...   
1282  Meet Fynn &amp; Taco. Fynn is an all-powerful ...   
1271  This is Eriq. His friend just reminded him of ...   
1967  This is Kial. Kial is either wearing a cape, w...   
1976  Two dogs in this one. Both are rare Jujitsu Py...   
1883  Here we have Pancho and Peaches. Pancho is a C...   
1920  This is Spark. He's nervous. Other dog hasn't ...   
2009  These are Peruvian Feldspars. Their names are ...   
2038  This is an Albanian 3 1/2 legged  Episcopalian...   
1718  10/10 for dog. 7/10 for cat. 12/10 for human. ...   
1820  Meet Hank and Sully. Hank is very proud of the...   
1771  Meet Holly. She's trying to teach small human-...   
1221  Meet Oliviér. He takes killer selfies. Has a d...   
863   After so many requests, this is Bretagne. She ...   
988   This is Bluebert. He just saw that both #Final...   
1004  Meet Travis and Flurp. Travis is pretty chill ...   
954   Happy 4/20 from the squad! 13/10 for all https...   
605   "Yep... just as I suspected. You're not flossi...   
812   This is Bookstore and Seaweed. Bookstore is ti...   
5796  Meet Eve. She's a raging alcoholic 8/10 (would...   
5727  Meet Rufio. He is unaware of the pink legless ...   

                                          expanded_urls rating_numerator  \
1392  https://twitter.com/dog_rates/status/682962037...               10   
1613  https://twitter.com/dog_rates/status/674646392...                8   
1513  https://twitter.com/dog_rate

#### Define
Fix rating numerator that have decimals.

#### Code


In [55]:

# View tweets with decimals in rating in 'text' column
twitter_archive_clean[twitter_archive_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]

C:\Users\topp\Anaconda3\lib\site-packages\pandas\core\strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


tweet_id                  timestamp              source  \
1438  680494726643068929  2015-12-25 21:06:00 +0000  Twitter for iPhone   
548   786709082849828864  2016-10-13 23:23:56 +0000  Twitter for iPhone   
40    883482846933004288  2017-07-08 00:28:19 +0000  Twitter for iPhone   
4721  778027034220126208  2016-09-20 00:24:34 +0000  Twitter for iPhone   

                                                   text  \
1438  Here we have uncovered an entire battalion of ...   
548   This is Logan, the Chow who lived. He solemnly...   
40    This is Bella. She hopes her smile made you sm...   
4721  This is Sophie. She's a Jubilant Bush Pupper. ...   

                                          expanded_urls rating_numerator  \
1438  https://twitter.com/dog_rates/status/680494726...               26   
548   https://twitter.com/dog_rates/status/786709082...               75   
40    https://twitter.com/dog_rates/status/883482846...                5   
4721  https://twitter.com/dog_rates/status/778027034...               27   

      rating_denominator    name  favorites  retweets  user_followers  \
1438                  10    None       1689       477         8822123   
548                   10   Logan      18383      6171         8822101   
40                    10   Bella      42564      8900         8822079   
4721                  10  Sophie       6574      1617         8822102   

                                              jpg_url dog_stage  \
1438  https://pbs.twimg.com/media/CXGaVxOWAAADjhF.jpg      None   
548   https://pbs.twimg.com/media/CurzvFTXgAA2_AP.jpg      None   
40    https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg      None   
4721  https://pbs.twimg.com/media/Cswbc2yWcAAVsCJ.jpg    pupper   

     prediction_algorithm  confidence_level  
1438               kuvasz          0.438627  
548            Pomeranian          0.467321  
40       golden_retriever          0.943082  
4721              clumber          0.946718

In [56]:
# Set correct numerators for specific tweets
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 883482846933004288) & (twitter_archive_clean['rating_numerator'] == 5), ['rating_numerator']] = 13.5
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 786709082849828864) & (twitter_archive_clean['rating_numerator'] == 75), ['rating_numerator']] = 9.75
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 778027034220126208) & (twitter_archive_clean['rating_numerator'] == 27), ['rating_numerator']] = 11.27
twitter_archive_clean.loc[(twitter_archive_clean['tweet_id'] == 680494726643068929) & (twitter_archive_clean['rating_numerator'] == 26), ['rating_numerator']] = 11.26

#### Test

In [57]:
twitter_archive_clean[twitter_archive_clean.text.str.contains(r"(\d+\.\d*\/\d+)")]


C:\Users\topp\Anaconda3\lib\site-packages\pandas\core\strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


tweet_id                  timestamp              source  \
1438  680494726643068929  2015-12-25 21:06:00 +0000  Twitter for iPhone   
548   786709082849828864  2016-10-13 23:23:56 +0000  Twitter for iPhone   
40    883482846933004288  2017-07-08 00:28:19 +0000  Twitter for iPhone   
4721  778027034220126208  2016-09-20 00:24:34 +0000  Twitter for iPhone   

                                                   text  \
1438  Here we have uncovered an entire battalion of ...   
548   This is Logan, the Chow who lived. He solemnly...   
40    This is Bella. She hopes her smile made you sm...   
4721  This is Sophie. She's a Jubilant Bush Pupper. ...   

                                          expanded_urls rating_numerator  \
1438  https://twitter.com/dog_rates/status/680494726...            11.26   
548   https://twitter.com/dog_rates/status/786709082...             9.75   
40    https://twitter.com/dog_rates/status/883482846...             13.5   
4721  https://twitter.com/dog_rates/status/778027034...            11.27   

      rating_denominator    name  favorites  retweets  user_followers  \
1438                  10    None       1689       477         8822123   
548                   10   Logan      18383      6171         8822101   
40                    10   Bella      42564      8900         8822079   
4721                  10  Sophie       6574      1617         8822102   

                                              jpg_url dog_stage  \
1438  https://pbs.twimg.com/media/CXGaVxOWAAADjhF.jpg      None   
548   https://pbs.twimg.com/media/CurzvFTXgAA2_AP.jpg      None   
40    https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg      None   
4721  https://pbs.twimg.com/media/Cswbc2yWcAAVsCJ.jpg    pupper   

     prediction_algorithm  confidence_level  
1438               kuvasz          0.438627  
548            Pomeranian          0.467321  
40       golden_retriever          0.943082  
4721              clumber          0.946718

#### Define
Get Dogs gender column from text column

#### Code

In [58]:
# Loop on all the texts and check if it has one of pronouns of male or female
# and append the result in a list

male = ['He', 'he', 'him', 'his', "he's", 'himself']
female = ['She', 'she', 'her', 'hers', 'herself', "she's"]

dog_gender = []

for text in twitter_archive_clean['text']:
    # Male
    if any(map(lambda v:v in male, text.split())):
        dog_gender.append('male')
    # Female
    elif any(map(lambda v:v in female, text.split())):
        dog_gender.append('female')
    # If group or not specified
    else:
        dog_gender.append('NaN')

# Test
len(dog_gender)

# Save the result in a new column 'dog_name'
twitter_archive_clean['dog_gender'] = dog_gender

#### Test

In [59]:
print("dog_gender count \n", twitter_archive_clean.dog_gender.value_counts())

dog_gender count 
 NaN       1130
male       633
female     224
Name: dog_gender, dtype: int64


#### Define
Convert the null values to None type

#### Code

In [60]:
twitter_archive_clean.loc[twitter_archive_clean['prediction_algorithm'] == 'NaN', 'prediction_algorithm'] = None
twitter_archive_clean.loc[twitter_archive_clean['dog_gender'] == 'NaN', 'dog_gender'] = None
twitter_archive_clean.loc[twitter_archive_clean['rating_numerator'] == 'NaN', 'rating_numerator'] = 0
#twitter_archive_clean.loc[twitter_archive_clean['rating_denominator'] == 'NaN', 'rating_denominator'] = 0

#### Test

In [61]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1987 entries, 0 to 7053
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   tweet_id              1987 non-null   int64  
 1   timestamp             1987 non-null   object 
 2   source                1987 non-null   object 
 3   text                  1987 non-null   object 
 4   expanded_urls         1987 non-null   object 
 5   rating_numerator      1987 non-null   object 
 6   rating_denominator    1987 non-null   int64  
 7   name                  1987 non-null   object 
 8   favorites             1987 non-null   int64  
 9   retweets              1987 non-null   int64  
 10  user_followers        1987 non-null   int64  
 11  jpg_url               1987 non-null   object 
 12  dog_stage             1987 non-null   object 
 13  prediction_algorithm  1679 non-null   object 
 14  confidence_level      1987 non-null   float64
 15  dog_gender           

#### Define
Change datatypes .

#### Code

In [62]:
twitter_archive_clean['tweet_id'] = twitter_archive_clean['tweet_id'].astype(str)
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean.timestamp)
twitter_archive_clean['source'] = twitter_archive_clean['source'].astype('category')
twitter_archive_clean['favorites'] = twitter_archive_clean['favorites'].astype(int)
twitter_archive_clean['retweets'] = twitter_archive_clean['retweets'].astype(int)
twitter_archive_clean['user_followers'] = twitter_archive_clean['user_followers'].astype(int)
twitter_archive_clean['dog_stage'] = twitter_archive_clean['dog_stage'].astype('category')
twitter_archive_clean['rating_numerator'] = twitter_archive_clean['rating_numerator'].astype(float)
twitter_archive_clean['rating_denominator'] = twitter_archive_clean['rating_denominator'].astype(float)
twitter_archive_clean['dog_gender'] = twitter_archive_clean['dog_gender'].astype('category')


####  Test

In [63]:
twitter_archive_clean.dtypes

tweet_id                             object
timestamp               datetime64[ns, UTC]
source                             category
text                                 object
expanded_urls                        object
rating_numerator                    float64
rating_denominator                  float64
name                                 object
favorites                             int32
retweets                              int32
user_followers                        int32
jpg_url                              object
dog_stage                          category
prediction_algorithm                 object
confidence_level                    float64
dog_gender                         category
dtype: object

#### Store

In [64]:
# Save clean DataFrame to csv file
twitter_archive_clean.drop(twitter_archive_clean.columns[twitter_archive_clean.columns.str.contains('Unnamed',case = False)],axis = 1)
twitter_archive_clean.to_csv('twitter_archive_master.csv', encoding = 'utf-8', index=False)

In [65]:
twitter_archive_clean = pd.read_csv('twitter_archive_master.csv')
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1987 entries, 0 to 1986
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   tweet_id              1987 non-null   int64  
 1   timestamp             1987 non-null   object 
 2   source                1987 non-null   object 
 3   text                  1987 non-null   object 
 4   expanded_urls         1987 non-null   object 
 5   rating_numerator      1987 non-null   float64
 6   rating_denominator    1987 non-null   float64
 7   name                  1987 non-null   object 
 8   favorites             1987 non-null   int64  
 9   retweets              1987 non-null   int64  
 10  user_followers        1987 non-null   int64  
 11  jpg_url               1987 non-null   object 
 12  dog_stage             1987 non-null   object 
 13  prediction_algorithm  1679 non-null   object 
 14  confidence_level      1987 non-null   float64
 15  dog_gender           